pkg/resource/deploy/step_generator.go

```diff
@@ -690,20 +694,11 @@ func (sg *stepGenerator) diff(urn resource.URN, old, new *resource.State, oldInp
 		return plugin.DiffResult{Changes: plugin.DiffSome, ReplaceKeys: []resource.PropertyKey{"provider"}}, nil
 	}
 
-	// Workaround #1251: unexpected replaces.
```

Comment:
Since we are dialing this back, it feels like we could land back in these states again.  I'm trying to understand what the guidance we should be capturing here is for end users.  I'm guessing that with these changes, it may happen that when they update their CLI to a version that includes this change, they will now see these spurious diffs.  I have a few questions:

1. Do these go away when providers start returning detailed information about a diff (I think the answer here is no, but at least our output will now make it clearer as to what is causing the diff.
2. Are these due to bugs in the providers that we would then go fix?  (I think the answer is yes, but I am not sure if there are classes of providers where the answer is no for some reason).
3. Is the guidance we provider to folks "use ignoreChanges to work around these bugs" or something else?

I think we should add some clear messaging to the CHANGELOG.md here, and I would love to understand the different experiences for end users when using a CLI with these changes and say a tf-bridged provider that does not return detailed diffs vs one that does.

Reply:
> Do these go away when providers start returning detailed information about a diff (I think the answer here is no, but at least our output will now make it clearer as to what is causing the diff.

Your analysis is correct. These do not go away, but the output should clearly indicate that the provider believes that there is a diff.

> Are these due to bugs in the providers that we would then go fix? (I think the answer is yes, but I am not sure if there are classes of providers where the answer is no for some reason).

Yes. My intuition is that the most common class of bug here is TF schema problems (esp. non-`Computed` fields that are written to by `Read`).

> Is the guidance we provider to folks "use ignoreChanges to work around these bugs" or something else?

`ignoreChanges` will not work for these cases until we have plumbed it through the provider interface. I think that the guidance will have to be "file a bug", unfortunately. We could hold off on this change until we do that work, though doing so will cause some minor complications for the import feature.

Reply:
I wonder if it would be worthwhile to have something like `PULUMI_USE_LEGACY_DIFFING_BEHAVIOR` that you could set to get back to the older behavior, and so we have a story for folks to unblock themselves if needed, and then remove that after we've updated everything and feel comfortable about having a satisfying story here.  I believe Sean did a similar thing for some of the plugin versioning changes we made.  Just a thought.

Reply:
Do you think that we need a more precise tool? I can imagine doing something like adding a `useLegacyDiff` property to `ResourceOptions`. That feels a little bit hacky, but would at least let users control this on a per-resource granularity.

Reply:
I went ahead with the envvar approach, though I named it `PULUMI_ENABLE_LEGACY_DIFF`.

Reply:
> Do you think that we need a more precise tool? I can imagine doing something like adding a useLegacyDiff property to ResourceOptions. That feels a little bit hacky, but would at least let users control this on a per-resource granularity.

My gut says no, but I defer to you here.  My thought about the environment variable was based by the following thoughts (which may or may not be correct):

1. Because we require changes across pulumi, the tf providers and the kubernetes provider, it's possible that we'll have a period of time where not everything is upgraded and things behave "strangely" (or everything does get upgraded but we discover some bugs which take us some time to address).  In this case being able to say: "I want to go back to the way things were before these changes" in a very blunt fashion seems useful.
2. We believe the new behavior is the correct long term decision for the product and that's the behavior we want long term.

If (2) is correct (and I think it is) then providing a first class way to opt out of this stuff on a per resource basis isn't something we would want to support long term.  So I'd rather not force folks add properties to ResourceOptions that we know we'd be deprecating and removing shortly.  Having this controlled outside the program and in a blunt fashion feels better to me because it makes it clearer that we believe the default behavior is correct and if you are opting out of it its because there's a bug you have to work around while we fix it.

I do see the appeal of having something more targeted here, but I worry about pushing something into the programming model in a more first class way that allows you to do something that long term we believe is the wrong behavior.

I think that you have a better intuition as to how this is gonna go, so I trust whatever you decide.